src/Compilers/CSharp/Test/Emit/Emit/DeterministicTests.cs

```diff
@@ -184,6 +185,83 @@ public void TestWriteOnlyStream()
             compilation.Emit(output);
         }
 
+        [Fact, WorkItem(11990, "https://github.com/dotnet/roslyn/issues/11990")]
+        public void ForwardedTypesAreEmittedInADeterministicOrder()
+        {
+            var forwardedToCode = @"
+namespace Namespace2 {
+    public class GenericType1<T> {}
+    public class GenericType3<T> {}
+    public class GenericType2<T> {}
+}
+namespace Namespace1 {
+    public class Type3 {}
+    public class Type2 {}
+    public class Type1 {}
+}
+namespace Namespace3 {
+    public class GenericType2<T, U> {}
+    public class GenericType1<T, U> {}
+}
+namespace Namespace4 {
+    namespace Embedded {
+        public class Type2 {}
+        public class Type1 {}
+    }
+}
+";
+            var forwardedToCompilation = CreateCompilation(forwardedToCode);
+            var forwardedToReference = new CSharpCompilationReference(forwardedToCompilation);
+
+            var forwardingCode = @"
+using System.Runtime.CompilerServices;
+[assembly: TypeForwardedTo(typeof(Namespace2.GenericType1<int>))]
+[assembly: TypeForwardedTo(typeof(Namespace2.GenericType3<int>))]
+[assembly: TypeForwardedTo(typeof(Namespace2.GenericType2<int>))]
+[assembly: TypeForwardedTo(typeof(Namespace1.Type3))]
+[assembly: TypeForwardedTo(typeof(Namespace1.Type2))]
+[assembly: TypeForwardedTo(typeof(Namespace1.Type1))]
+[assembly: TypeForwardedTo(typeof(Namespace3.GenericType2<int, int>))]
+[assembly: TypeForwardedTo(typeof(Namespace3.GenericType1<int, int>))]
+[assembly: TypeForwardedTo(typeof(Namespace4.Embedded.Type2))]
+[assembly: TypeForwardedTo(typeof(Namespace4.Embedded.Type1))]
+";
+
+            var forwardingCompilation = CreateCompilationWithMscorlib(forwardingCode, new MetadataReference[] { forwardedToReference });
+            var metadataFullNames = new List<string>();
+
+            using (var stream = forwardingCompilation.EmitToStream())
+            {
+                using (var block = ModuleMetadata.CreateFromStream(stream))
+                {
+                    foreach (var handle in block.MetadataReader.ExportedTypes)
+                    {
+                        var type = block.MetadataReader.GetExportedType(handle);
+                        var typeName = block.MetadataReader.GetString(type.Name);
+                        var namespaceName = block.MetadataReader.GetString(type.Namespace);
+
+                        metadataFullNames.Add($"{namespaceName}.{typeName}");
+                    }
```

Comment:
Minor point: Consider extracting a helper method shared with VB. See `MetadataValidation.GetFullTypeNames()` for example.